### ISSUE TITLE
代码与论文不一致

### ISSUE THREAD

**Issue by LMissher** [DAMO-DI-ML/KDD2023-DCdetector#16]
很棒的工作！

通过阅读论文与代码，可以看出论文复用了AnomalyTransformer的训练框架，但是忽略了与其不一样的地方，也就导致本文代码与论文不一致。

1. Training loss
代码里168行 loss = prior_loss - series_loss，其中 prior_loss = KL(P, Stopgrad(N))  + KL(Stopgrad(N), P)，series_loss = KL(N, Stopgrad(P))  + KL(Stopgrad(P), N)，也就有 loss = KL(P, Stopgrad(N))  + KL(Stopgrad(N), P) - ( KL(N, Stopgrad(P))  + KL(Stopgrad(P), N) )，这与论文中 loss = 1/2*KL(P, Stopgrad(N)) + 1/2*KL(N, Stopgrad(P)) 不一致。这是因为 AnomalyTransformer 采用了Minmax训练策略，而本文并没有。此外 loss 为 0 是因为数值上 prior_loss 与 series_loss 是一样的，只是梯度停止的变量不一样，所以训练是有效的。希望后续能把代码更新到与论文一致，或者把arxiv上的论文更新为与代码一致，避免对大家带来误导。

2. Test metric
代码里223、255、290行 metric = torch.softmax((-series_loss - prior_loss), dim=-1) 应该为 metric = torch.softmax((series_loss + prior_loss), dim=-1)。这是因为 AnomalyTransformer 中两个分布的 KL 散度差异越小越异常，而本文中两个分布的 KL 散度差异越大越异常，也即本文中异常点的 KL 散度应该更大。从根本上说是本文中两个正常表示的分布更一致，而 AT 是异常点与先验分布一致。我实验后确实是加法效果更好点。希望后续能修改错误的 metric 代码。

**Comment by tianzhou2011** [comment]
谢谢您很细致复现了文章，确实这里有和代码不一致的地方，我们会尽快更新一版arxiv的版本，避免误解

**Comment by LMissher** [comment]
期待后续的更新~
